File: rustbridge-python/rustbridge/core/response_envelope.py
"""Response envelope for parsing JSON responses from FFI."""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any

from rustbridge.core.plugin_exception import PluginException


@dataclass
class ResponseEnvelope:
    """
    Response envelope wrapping a response from FFI transport.

    Attributes:
        status: Response status ("success" or "error").
        payload: Response payload (on success) or None.
        error_code: Error code (on failure) or None.
        error_message: Error message (on failure) or None.
        request_id: Original request ID for correlation (optional).
    """

    status: str
    payload: Any | None = None
    error_code: int | None = None
    error_message: str | None = None
    request_id: int | None = None

    @property
    def is_success(self) -> bool:
        """Check if this is a success response."""
        return self.status == "success"

    @classmethod
    def from_json(cls, json_str: str) -> ResponseEnvelope:
        """
        Parse a ResponseEnvelope from JSON string.

        Args:
            json_str: The JSON string.

        Returns:
            The parsed ResponseEnvelope.

        Raises:
            PluginException: If parsing fails.
        """
        try:
            data = json.loads(json_str)
        except json.JSONDecodeError as e:
            raise PluginException(f"Failed to parse response JSON: {e}") from e

        return cls(
            status=data.get("status", "error"),
            payload=data.get("payload"),
            error_code=data.get("error_code"),
            error_message=data.get("error_message"),
            request_id=data.get("request_id"),
        )

    @classmethod
    def from_bytes(cls, data: bytes) -> ResponseEnvelope:
        """
        Parse a ResponseEnvelope from bytes.

        Args:
            data: The JSON bytes.

        Returns:
            The parsed ResponseEnvelope.
        """
        return cls.from_json(data.decode("utf-8"))

    def get_payload_json(self) -> str:
        """
        Get the payload as a JSON string.

        Returns:
            The payload serialized as JSON, or "null" if no payload.
        """
        if self.payload is None:
            return "null"
        return json.dumps(self.payload)

    def to_exception(self) -> PluginException:
        """
        Convert this error response to a PluginException.

        Returns:
            A PluginException with the error details.
        """
        return PluginException(
            message=self.error_message or "Unknown error",
            error_code=self.error_code or 1,
        )

    def unwrap(self) -> Any:
        """
        Unwrap the payload, raising an exception if this is an error response.

        Returns:
            The payload value.

        Raises:
            PluginException: If this is an error response.
        """
        if not self.is_success:
            raise self.to_exception()
        return self.payload

File: rustbridge-csharp/RustBridge.Core/ResponseEnvelope.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace RustBridge;

/// <summary>
/// Response envelope from plugin calls.
/// <para>
/// All plugin responses are wrapped in this envelope format, which includes
/// the type tag, payload, and optional error information.
/// </para>
/// </summary>
public class ResponseEnvelope
{
    /// <summary>
    /// The message type tag.
    /// </summary>
    [JsonPropertyName("type_tag")]
    public string? TypeTag { get; set; }

    /// <summary>
    /// The response payload as a JSON element.
    /// </summary>
    [JsonPropertyName("payload")]
    public JsonElement? Payload { get; set; }

    /// <summary>
    /// The error code (0 = success).
    /// </summary>
    [JsonPropertyName("error_code")]
    public int ErrorCode { get; set; }

    /// <summary>
    /// The error message (if any).
    /// </summary>
    [JsonPropertyName("error_message")]
    public string? ErrorMessage { get; set; }

    /// <summary>
    /// Check if this response represents a success.
    /// </summary>
    public bool IsSuccess => ErrorCode == 0;

    /// <summary>
    /// Get the payload as a JSON string.
    /// </summary>
    /// <returns>The payload JSON string, or "null" if no payload.</returns>
    public string GetPayloadJson()
    {
        if (Payload is null)
        {
            return "null";
        }
        return Payload.Value.GetRawText();
    }

    /// <summary>
    /// Convert this error response to a PluginException.
    /// </summary>
    /// <returns>A PluginException with the error details.</returns>
    public PluginException ToException()
    {
        return new PluginException(ErrorCode, ErrorMessage ?? "Unknown error");
    }

    /// <summary>
    /// Parse a response envelope from JSON.
    /// </summary>
    /// <param name="json">The JSON string.</param>
    /// <returns>The parsed envelope.</returns>
    public static ResponseEnvelope FromJson(string json)
    {
        return JsonSerializer.Deserialize<ResponseEnvelope>(json)
            ?? throw new PluginException("Failed to parse response envelope");
    }
}
